Add delimiter support for spatial dimensions. (#3690)

jihoonson opened the pull request:
This is not a complete patch, but I would like to check this is a right direction.
Spatial dimensions work well with TSV format, but is not with CSV.
I think the root cause is the fixed delimiter for spatial dimensions which is same with that of CSV, and added a configuration to set a delimiter for spatial dimensions.
If this is a right direction, I'll add some tests and documents.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/druid-io/druid/3718)
<!-- Reviewable:end -->

drcrallen commented:
https://github.com/druid-io/druid/issues/3690 related

navis commented on api/src/main/java/io/druid/data/input/impl/NewSpatialDimensionSchema.java:
```diff
@@ -51,6 +55,11 @@ public NewSpatialDimensionSchema(
   }
 
+  @JsonProperty
+  public String getDelimiter() {
```
format. I can see some more in below

jihoonson replied:
Fixed. Thanks.

navis commented:
test cases with custom delimiter?

leventov commented on api/src/main/java/io/druid/data/input/impl/NewSpatialDimensionSchema.java:
```diff
@@ -76,13 +85,20 @@ public boolean equals(Object o)
     NewSpatialDimensionSchema that = (NewSpatialDimensionSchema) o;
 
-    return dims != null ? dims.equals(that.dims) : that.dims == null;
+    if (dims != null ? dims.equals(that.dims) : that.dims == null) {
```
It's `java.util.Objects.equals(dims, that.dims)`

jihoonson replied:
Fixed. Thanks.

leventov commented on api/src/main/java/io/druid/data/input/impl/SpatialDimensionSchema.java:
```diff
@@ -67,15 +75,20 @@ public boolean equals(Object o)
     }
-    return dims != null ? dims.equals(that.dims) : that.dims == null;
 
+    if (dims != null ? !dims.equals(that.dims) : that.dims != null) {
```
It's `java.util.Objects.equals(dims, that.dims)`

leventov commented on processing/src/main/java/io/druid/query/filter/SpatialDimFilter.java:
```diff
@@ -33,11 +34,13 @@
 public class SpatialDimFilter implements DimFilter
 {
   private final String dimension;
+  private final String delimiter;
```
Shouldn't it be part of `equals()`, `hashCode()` and `toString()` of this class?

jihoonson replied:
Fixed. Thanks.

leventov commented on processing/src/main/java/io/druid/segment/incremental/SpatialDimensionRowTransformer.java:
```diff
@@ -46,19 +46,27 @@
 {
-  private static final Joiner JOINER = Joiner.on(",");
-  private static final Splitter SPLITTER = Splitter.on(",");
+  private final Map<String, Joiner> joinerCache;
```
Maybe encapsulate joiner and splitter in `SpatialDimensionSchema` and add methods like `SpatialDimensionSchema.isJoinedValueValid()` and another one, whose implementation is `joiner.join(spatialDimVals)`? It would allow to avoid making those maps, populating and querying them.

leventov commented on processing/src/test/java/io/druid/segment/filter/SpatialFilterBonusTest.java:
```diff
@@ -306,6 +309,7 @@ private static QueryableIndex makeMergedQueryableIndex(final IndexSpec indexSpec
                                                           new SpatialDimensionSchema(
                                                               "dim.geo",
                                                               Lists.<String>newArrayList()
+                                                              ,null
```
formatting

jihoonson replied:
Fixed.

jihoonson commented:
@navis @leventov thanks for your comment. I added some tests. 
I also added support for custom spatial dimension delimiter to ```IndexSpec``` for ```IndexMerger```.

leventov commented on processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java:
```diff
@@ -73,7 +74,7 @@
   protected static final ListIndexed EMPTY_STR_DIM_VAL = new ListIndexed<>(Arrays.asList(""), String.class);
   protected static final int[] EMPTY_STR_DIM_ARRAY = new int[]{0};
-  protected static final Splitter SPLITTER = Splitter.on(",");
+//  protected static final Splitter SPLITTER = Splitter.on(",");
```
Please remove this line

jihoonson replied:
Removed. Thanks.

leventov commented on processing/src/main/java/io/druid/query/filter/SpatialDimFilter.java:
```diff
@@ -111,14 +122,15 @@ public boolean equals(Object o)
     }
 
-    return true;
+    return Objects.equal(delimiter, that.delimiter);
```
delimiter is non-null, so it could be just delimiter.equals(that.delimiter)

leventov replied:
Also this needlessly uses Guava's utility class and methods, which are deprecated since Java 7 in favor of java.utils.Objects.

leventov commented on processing/src/main/java/io/druid/segment/IndexSpec.java:
```diff
@@ -156,6 +168,11 @@ public boolean equals(Object o)
       return false;
     }
+
+    if (!Objects.equals(spatialDimensionDelimiter, indexSpec.spatialDimensionDelimiter)) {
```
spatialDimensionDelimiter is non-null, this line could be simplified

leventov commented on processing/src/main/java/io/druid/segment/IndexSpec.java:
```diff
@@ -167,6 +184,7 @@ public int hashCode()
     int result = bitmapSerdeFactory != null ? bitmapSerdeFactory.hashCode() : 0;
     result = 31 * result + (dimensionCompression != null ? dimensionCompression.hashCode() : 0);
     result = 31 * result + (metricCompression != null ? metricCompression.hashCode() : 0);
+    result = 31 * result + (spatialDimensionDelimiter != null ? spatialDimensionDelimiter.hashCode() : 0);
```
spatialDimensionDelimiter is non-null, this line could be simplified

leventov commented on api/src/main/java/io/druid/data/input/impl/SpatialDimensionSchema.java:
```diff
@@ -67,15 +96,20 @@ public boolean equals(Object o)
     if (dimName != null ? !dimName.equals(that.dimName) : that.dimName != null) {
```
Using two different styles of comparison of nullable objects in one method could raise questions, could you please also replace this with Objects.equals()?

jihoonson commented:
I simplified equals() and hashCode(), and fixed additional bugs in them.

leventov commented on api/src/main/java/io/druid/data/input/impl/NewSpatialDimensionSchema.java:
```diff
@@ -76,13 +105,24 @@ public boolean equals(Object o)
     NewSpatialDimensionSchema that = (NewSpatialDimensionSchema) o;
 
-    return dims != null ? dims.equals(that.dims) : that.dims == null;
+    if (!super.equals(that)) {
```
I've never seen `super.equals()` call in `equals()` implementations. This is fragile, e. g. if `DimensionSchema` changed and instead of `getClass() != that.getClass()` checks `DimensionSchema.class != that.getClass()`, which is also a popular pattern in `equals()` implementations (although it would probably be wrong for DimensionSchema, if it is designed for subclassing).

So I suggest either to add a test for `NewSpatialDimensionSchema.equals()`, so if `DimensionSchema` is changed unintentionally, the test will catch this error. Or, maybe, replace subclassing with composition.

jihoonson replied:
Calling ```super.equals()``` might be an old style, but is still used even in Druid itself (https://github.com/druid-io/druid/blob/master/processing/src/main/java/io/druid/query/select/SelectQuery.java#L236), or Guava (https://github.com/google/guava/blob/master/guava/src/com/google/common/primitives/Longs.java#L678). Maybe we can remove ```super.equals()``` later if we can avoid evaluating equality or do some refactoring. 

I think replacing subclassing with composition is a refactoring issue and is not covered by this ticket, so added a simple test.

leventov commented:
@jihoonson do you still need this?

jihoonson commented:
@leventov hmm, no. I'll close this pr.
